src/pharmpy/internals/fs/tmp.py

```diff
@@ -1,83 +1,104 @@
-        """
+        with TemporaryDirectory() as tmpdir:
+            ...
 
-        def __init__(self, suffix=None, prefix=None, dir=None):
-            self.name = mkdtemp(suffix, prefix, dir)
-            self._finalizer = weakref.finalize(
-                self,
-                self._cleanup,
-                self.name,
-                warn_message="Implicitly cleaning up {!r}".format(self),
-            )
+    Upon exiting the context, the directory and everything contained
+    in it are removed.
+    """
 
-        @classmethod
-        def _rmtree(cls, name):
-            def onerror(func, path, exc_info):
-                if issubclass(exc_info[0], PermissionError):
+    def __init__(self, suffix=None, prefix=None, dir=None, ignore_cleanup_errors=False):
+        self.name = mkdtemp(suffix, prefix, dir)
+        self._ignore_cleanup_errors = ignore_cleanup_errors
+        self._finalizer = _weakref.finalize(
+            self,
+            self._cleanup,
+            self.name,
+            warn_message="Implicitly cleaning up {!r}".format(self),
+            ignore_errors=self._ignore_cleanup_errors,
+        )
 
-                    def resetperms(path):
-                        try:
-                            os.chflags(path, 0)
-                        except AttributeError:
-                            pass
-                        os.chmod(path, 0o700)
+    @classmethod
+    def _rmtree(cls, name, ignore_errors=False):
+        def onerror(func, path, exc_info):
+            if issubclass(exc_info[0], PermissionError):
 
+                def resetperms(path):
                     try:
-                        if path != name:
-                            resetperms(os.path.dirname(path))
-                        resetperms(path)
+                        _os.chflags(path, 0)
+                    except AttributeError:
+                        pass
+                    _os.chmod(path, 0o700)
 
-                        try:
-                            os.unlink(path)
+                try:
+                    if path != name:
+                        resetperms(_os.path.dirname(path))
+                    resetperms(path)
+
+                    try:
+                        _os.unlink(path)
+                    except IsADirectoryError:
+                        cls._rmtree(path, ignore_errors=ignore_errors)
+                    except PermissionError as pe:
                         # PermissionError is raised on FreeBSD for directories
-                        except (IsADirectoryError, PermissionError):
-                            time.sleep(0.1)
-                            cls._rmtree(path)
-                    except FileNotFoundError:
-                        pass
-                elif issubclass(exc_info[0], FileNotFoundError):
+                        # and by Windows on lock files used by other processes
+                        try:
+                            cls._rmtree(path, ignore_errors=ignore_errors)
```

Comment:
This somehow assumes a `PermissionError` will not be raised again, which is bold, since it can potentially recurse forever. This is not a problem we have encountered, and it is taken straight from `cpython`'s source so I did not change it.